File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    sort(times.begin(), times.end());

    long long left = 1;
    long long right = (long long)times.back() * (long long)n; // 최악의 경우
    long long answer = right;

    while (left <= right) {
        long long mid = left + (right - left) / 2;

        long long processed = 0;
        for (int t : times) {
            processed += mid / (long long)t;
            if (processed >= n) break; // 충분하면 더 안 더해도 됨
        }

        if (processed >= n) {
            answer = mid;
            right = mid - 1; // 더 작은 시간 가능?
        } else {
            left = mid + 1;  // 시간이 더 필요
        }
    }

    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.Arrays;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);

        long left = 1L;
        long right = (long) times[times.length - 1] * (long) n; // 최악의 경우
        long answer = right;

        while (left <= right) {
            long mid = left + (right - left) / 2;

            long processed = 0L;
            for (int t : times) {
                processed += mid / (long) t;
                if (processed >= n) break; // 충분하면 더 계산 X
            }

            if (processed >= n) {
                answer = mid;
                right = mid - 1; // 더 작은 시간 탐색
            } else {
                left = mid + 1;  // 더 큰 시간 필요
            }
        }

        return answer;
    }
}